File: cpp/merge_k_sorted_lists.cpp
23. Merge k Sorted Lists Add to List
DescriptionSubmissionsSolutions
Total Accepted: 138657
Total Submissions: 522055
Difficulty: Hard
Contributor: LeetCode
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.

Subscribe to see which companies asked this question.

Hide Tags Divide and Conquer Linked List Heap
Hide Similar Problems (E) Merge Two Sorted Lists (M) Ugly Number II

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    class Compare {
    public:
        bool operator()(ListNode* l1, ListNode* l2) {
            return l1->val>l2->val;
        }
    };
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode dummy(0);
        ListNode* tail=&dummy;
        priority_queue<ListNode*, vector<ListNode*>, Compare> pq;
        for (auto l:lists) {
            if (l) pq.push(l);
        }
        while (!pq.empty()) {
            ListNode* t=pq.top();
            pq.pop();
            tail->next=t;
            tail=t;
            if (t->next) pq.push(t->next);
        }
        
        return dummy.next;
    }
};

File: python/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from Queue import PriorityQueue
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        dummy=tail=ListNode()
        pq=PriorityQueue()
        for l in lists:
            if l: pq.put((l.val, l))
        while not pq.empty():
            t=pq.get()[1]
            tail.next=t
            tail=t
            if t.next: pq.put((t.next.val, t.next))
        return dummy.next        


2. heapq
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        dummy=tail=ListNode()
        # in case of tie-break of val comparison, compare index instead of ListNode
        pq=[(l.val, i, l) for i, l in enumerate(lists) if l]
        heapify(pq)
        while pq:
            v, i, l=pq[0]
            heappop(pq)
            tail.next=l
            tail=l
            if l.next: heappush(pq, (l.next.val, i, l.next))
        return dummy.next

3.
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        dummy=tail=ListNode()
        # in case of tie-break of val comparison, compare index instead of ListNode
        pq=[]
        for i, l in enumerate(lists):
            if l:
                heappush(pq, (l.val, i, l))
        while pq:
            v, i, l=pq[0]
            heappop(pq)
            tail.next=l
            tail=l
            if l.next: heappush(pq, (l.next.val, i, l.next))
        return dummy.next

